### PR TITLE
Status-message consumer

### PR DESCRIPTION
**Pull request by tba76**
## Related Issue(s)
- #166 

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.Notifications.Core/Enums/EmailNotificationResultType.cs [hunk 1]
```diff
@@ -20,6 +20,11 @@ public enum EmailNotificationResultType
     /// </summary>
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Enums/EmailNotificationResultType.cs:20]: this description should be only sent, and not delivered as well

File: src/Altinn.Notifications.Core/Models/Notification/SendOperationResult.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    /// <summary>
+    /// The send operation id
+    /// </summary>
+    public string OperationId { get; set; } = string.Empty;
+
+    /// <summary>
+    /// The email send result
+    /// </summary>
+    public EmailNotificationResultType? SendResult { get; set; }
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Models/Notification/SendOperationResult.cs:25]: should this be nullable, or should we have a default value "Unknown"?

File: src/Altinn.Notifications.Core/Services/EmailNotificationService.cs [hunk 3]
```diff
@@ -71,6 +69,12 @@ public async Task SendNotifications()
         }
     }
 
+    /// <inheritdoc/>
+    public async Task UpdateStatusNotification(SendOperationResult sendOperationResult)
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Services/EmailNotificationService.cs:73]: updateNotificationsStatus maybe?

Comment (acn-sbuad) [reply src/Altinn.Notifications.Core/Services/EmailNotificationService.cs:73]: or UpdateSendStatus

File: src/Altinn.Notifications.Integrations/Kafka/Consumers/EmailStatusConsumer.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        _topicName = settings.Value.EmailStatusUpdatedTopicName;
+    }
+
+    /// <inheritdoc/>
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        return Task.Run(() => ConsumeMessage(ProcessOrder, RetryOrder, stoppingToken), stoppingToken);
+    }
+
+    private async Task ProcessOrder(string message)
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Integrations/Kafka/Consumers/EmailStatusConsumer.cs:41]: method name has not been changed.. ProcessStatus f.eks?

File: test/Altinn.Notifications.Tests/Notifications.Core/TestingServices/EmailNotificationServiceTests.cs [hunk 5]
```diff
@@ -136,6 +136,32 @@ public async Task CreateEmailNotification_ToAddressDefined_ResultFailedRecipient
         repoMock.Verify();
     }
 
+    [Fact]
+    public async Task UpdateStatusNotification()
```
Comment (acn-sbuad) [inline test/Altinn.Notifications.Tests/Notifications.Core/TestingServices/EmailNotificationServiceTests.cs:140]: Maybe update test name to match the new name of the service method

File: test/Altinn.Notifications.IntegrationTests/Notifications.Integrations/TestingConsumers/EmailStatusConsumerTests.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        // Act
+        await consumerService.StartAsync(CancellationToken.None);
+        await Task.Delay(10000);
+        await consumerService.StopAsync(CancellationToken.None);
+
+        // Assert
+
+        string emailNotificationStatus = await SelectEmailNotificationStatus(Notification.Id);
+        Assert.Equal(EmailNotificationResultType.Succeeded.ToString(), emailNotificationStatus);
+
```
Comment (acn-sbuad) [inline test/Altinn.Notifications.IntegrationTests/Notifications.Integrations/TestingConsumers/EmailStatusConsumerTests.cs:52]:
```suggestion
```

File: test/Altinn.Notifications.IntegrationTests/Notifications.Integrations/TestingConsumers/EmailStatusConsumerTests.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        await KafkaUtil.PublishMessageOnTopic(_statusUpdatedTopicName, sendOperationResult.Serialize());
+
+
+        // Act
+        await consumerService.StartAsync(CancellationToken.None);
+        await Task.Delay(10000);
+        await consumerService.StopAsync(CancellationToken.None);
+
+        // Assert
+
```
Comment (acn-sbuad) [inline test/Altinn.Notifications.IntegrationTests/Notifications.Integrations/TestingConsumers/EmailStatusConsumerTests.cs:49]:
```suggestion
```

File: src/Altinn.Notifications.Persistence/Repository/EmailNotificationRepository.cs [hunk 8]
```diff
@@ -19,7 +19,7 @@ public class EmailNotificationRepository : IEmailNotificationRepository
     private readonly NpgsqlDataSource _dataSource;
     private const string _insertEmailNotificationSql = "call notifications.insertemailnotification($1, $2, $3, $4, $5, $6, $7)"; // (__orderid, _alternateid, _recipientid, _toaddress, _result, _resulttime, _expirytime)
     private const string _getEmailNotificationsSql = "select * from notifications.getemails_statusnew_updatestatus()";
-    private const string _setResultStatus = "update notifications.emailnotifications set result =$1::emailnotificationresulttype where alternateid=$2";
+    private const string _setResultStatus = "call notifications.updateemailstatus($1, $2)"; // (_alternateid, _result)
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Persistence/Repository/EmailNotificationRepository.cs:22]: + operationId

File: src/Altinn.Notifications.Integrations/Kafka/Consumers/EmailStatusConsumer.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+namespace Altinn.Notifications.Integrations.Kafka.Consumers;
+
+/// <summary>
+/// Kafka consumer class for status messages about email notifications
+/// </summary>
+public class EmailStatusConsumer : KafkaConsumerBase<EmailStatusConsumer>
+{
+    private readonly IEmailNotificationService _emailNotificationsService;
+    private readonly IKafkaProducer _producer;
+    private readonly string _topicName;
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Integrations/Kafka/Consumers/EmailStatusConsumer.cs:18]: Do  we want to be clear that this is the retry topic name?

File: src/Altinn.Notifications.Integrations/Kafka/Consumers/EmailStatusConsumer.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        return Task.Run(() => ConsumeMessage(ProcessOrder, RetryOrder, stoppingToken), stoppingToken);
+    }
+
+    private async Task ProcessOrder(string message)
+    {
+        bool succeeded = SendOperationResult.TryParse(message, out SendOperationResult result);
+
+        if (!succeeded)
+        {
+            return;
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Integrations/Kafka/Consumers/EmailStatusConsumer.cs:47]: what I've done in notifications.email is add an error log here to inform that we had bad data in the queue. if not we wont pick up on this anywhere else

File: src/Altinn.Notifications.Persistence/Repository/EmailNotificationRepository.cs [hunk 11]
```diff
@@ -75,11 +75,12 @@ await using (NpgsqlDataReader reader = await pgcom.ExecuteReaderAsync())
     }
 
     /// <inheritdoc/>
-    public async Task SetResultStatus(Guid notificationId, EmailNotificationResultType status)
+    public async Task UpdateSendStatus(Guid notificationId, EmailNotificationResultType status, string? operationId)
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Persistence/Repository/EmailNotificationRepository.cs:78]: what if we set a default value null for operationId? making it obsolete to change existing use of the method ?

File: src/Altinn.Notifications.Persistence/Repository/EmailNotificationRepository.cs [hunk 12]
```diff
@@ -19,7 +19,7 @@ public class EmailNotificationRepository : IEmailNotificationRepository
     private readonly NpgsqlDataSource _dataSource;
     private const string _insertEmailNotificationSql = "call notifications.insertemailnotification($1, $2, $3, $4, $5, $6, $7)"; // (__orderid, _alternateid, _recipientid, _toaddress, _result, _resulttime, _expirytime)
     private const string _getEmailNotificationsSql = "select * from notifications.getemails_statusnew_updatestatus()";
-    private const string _setResultStatus = "update notifications.emailnotifications set result =$1::emailnotificationresulttype where alternateid=$2";
+    private const string _setResultStatus = "call notifications.updateemailstatus($1, $2, $3)"; // (_alternateid, _result, _operationid)
```
Comment (SandGrainOne) [inline src/Altinn.Notifications.Persistence/Repository/EmailNotificationRepository.cs:22]: Maybe also change the variable name.

File: test/Altinn.Notifications.Tests/Notifications.Core/TestingServices/EmailNotificationServiceTests.cs [hunk 13]
```diff
@@ -136,6 +136,32 @@ public async Task CreateEmailNotification_ToAddressDefined_ResultFailedRecipient
         repoMock.Verify();
     }
 
+    [Fact]
+    public async Task UpdateSendStatus()
```
Comment (SandGrainOne) [inline test/Altinn.Notifications.Tests/Notifications.Core/TestingServices/EmailNotificationServiceTests.cs:140]: I would like to see a slightly better test scenario description as the method name.

File: test/Altinn.Notifications.IntegrationTests/Notifications.Integrations/TestingConsumers/EmailStatusConsumerTests.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            { "KafkaSettings__EmailStatusUpdatedTopicName", _statusUpdatedTopicName },
+            { "KafkaSettings__Admin__TopicList", $"[\"{_statusUpdatedTopicName}\"]" }
+        };
+
+        using EmailStatusConsumer consumerService = (EmailStatusConsumer)ServiceUtil
+                                                    .GetServices(new List<Type>() { typeof(IHostedService) }, vars)
+                                                    .First(s => s.GetType() == typeof(EmailStatusConsumer))!;
+
+        (NotificationOrder Order, EmailNotification Notification) = await PostgreUtil.PopulateDBWithOrderAndEmailNotification(_sendersRef);
```
Comment (SandGrainOne) [inline test/Altinn.Notifications.IntegrationTests/Notifications.Integrations/TestingConsumers/EmailStatusConsumerTests.cs:32]: We can discard the values not in use.
```suggestion
        _ = await PostgreUtil.PopulateDBWithOrderAndEmailNotification(_sendersRef);
```

File: test/Altinn.Notifications.IntegrationTests/Notifications.Integrations/TestingConsumers/EmailStatusConsumerTests.cs [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        await consumerService.StartAsync(CancellationToken.None);
+        await Task.Delay(10000);
+        await consumerService.StopAsync(CancellationToken.None);
+
+        // Assert
+        string emailNotificationStatus = await SelectEmailNotificationStatus(Notification.Id);
+        Assert.Equal(EmailNotificationResultType.Succeeded.ToString(), emailNotificationStatus);
+    }
+
+    public async void Dispose()
```
Comment (SandGrainOne) [inline test/Altinn.Notifications.IntegrationTests/Notifications.Integrations/TestingConsumers/EmailStatusConsumerTests.cs:53]: I can't see any class fields holding on to any unmanaged resources. I prefer to use the IAsyncLifetime interface in place of IDisposable.

**Review by SandGrainOne — APPROVED** [review]

**Pull request merged**